cli/tools/upgrade.rs

```diff
@@ -15,12 +16,84 @@ use std::io::Write;
 use std::path::Path;
 use std::path::PathBuf;
 use std::process::Command;
+use std::time::Duration;
 
 static ARCHIVE_NAME: Lazy<String> =
   Lazy::new(|| format!("deno-{}.zip", env!("TARGET")));
 
 const RELEASE_URL: &str = "https://github.com/denoland/deno/releases";
 
+const UPGRADE_CHECK_INTERVAL: i64 = 12;
+
+const UPGRADE_CHECK_FETCH_DELAY: Duration = Duration::from_millis(500);
+
+pub fn check_for_upgrades(cache_dir: PathBuf) -> Option<String> {
+  if env::var("DENO_NO_UPDATE_CHECK").is_ok() {
+    return None;
+  }
+
+  let p = cache_dir.join("latest.txt");
+  let content = match std::fs::read_to_string(&p) {
+    Ok(file) => file,
+    Err(_) => "".to_string(),
+  };
+
+  let (mut last_checked, mut latest_version) = (None, None);
+  let split_content = content.split('!').collect::<Vec<_>>();
+  if split_content.len() == 2 {
+    last_checked = Some(split_content[0]);
+    latest_version = Some(split_content[1].to_owned());
+  }
+
+  if latest_version.is_none() || latest_version.as_ref().unwrap().is_empty() {
+    let last_checked_dt = last_checked.and_then(|last_checked| {
+      chrono::DateTime::parse_from_rfc3339(last_checked)
+        .map(|dt| dt.with_timezone(&chrono::Utc))
+        .ok()
+    });
+
+    let should_check = match last_checked_dt {
+      Some(last_checked_dt) => {
+        let last_check_age =
+          chrono::Utc::now().signed_duration_since(last_checked_dt);
+        last_check_age > chrono::Duration::hours(UPGRADE_CHECK_INTERVAL)
+      }
+      None => true,
+    };
+
+    if should_check {
+      tokio::spawn(async move {
+        // Sleep for a small amount of time to not unnecessarily impact startup
+        // time.
+        tokio::time::sleep(UPGRADE_CHECK_FETCH_DELAY).await;
+
+        // Fetch latest version or commit hash from server.
+        let client = match build_http_client(None) {
+          Ok(client) => client,
+          Err(_) => return,
+        };
+        let latest_version = match if version::is_canary() {
+          get_latest_canary_version(&client).await
+        } else {
+          get_latest_release_version(&client).await
+        } {
+          Ok(latest_version) => latest_version,
+          Err(_) => return,
+        };
+
+        // Write the latest version to `latest.txt`.
+        let contents =
+          format!("{}!{}", chrono::Utc::now().to_rfc3339(), latest_version);
+        let _ = std::fs::write(cache_dir.join("latest.txt"), contents);
```

Comment:
parsing and serialization of this file's contents should be separate functions with a unit test.